WeScan/Common/VisionRectangleDetector.swift

```diff
@@ -0,0 +1,65 @@
+//
+//  VisionRectangleDetector.swift
+//  WeScan
+//
+//  Created by Julian Schiavo on 28/7/2018.
+//  Copyright © 2018 WeTransfer. All rights reserved.
+//
+
+import Vision
+import Foundation
+
+/// Class used to detect rectangles from an image.
+@available(iOS 11.0, *)
+struct VisionRectangleDetector {
+    
+    /// Detects rectangles from the given image on iOS 11 and above.
+    ///
+    /// - Parameters:
+    ///   - image: The image to detect rectangles on.
+    /// - Returns: The biggest detected rectangle on the image.
+    static func rectangle(forImage image: CIImage, completion: @escaping ((Quadrilateral?) -> ())) {
+      
+        let imageRequestHandler = VNImageRequestHandler(ciImage: image, options: [:])
+        var biggestRectangle: Quadrilateral? = nil
+        
+        // Create the rectangle request, and, if found, return the biggest rectangle (else return nothing).
+        let rectangleDetectionRequest: VNDetectRectanglesRequest = {
+            let rectDetectRequest = VNDetectRectanglesRequest(completionHandler: { (request, error) in
+                if error == nil {
+                    guard let results = request.results as? [VNRectangleObservation] else { return }
+                
+                  //                    TODO: Fix all of this
+                  
+                    let quads: [Quadrilateral] = []
+
+                    guard let biggest = results.count > 1 ? quads.biggest() : quads.first else { return }
+                    
+                    let transform = CGAffineTransform.identity
+                        .scaledBy(x: image.extent.size.width, y: image.extent.size.height)
+                    
+                    biggestRectangle = Quadrilateral(topLeft: biggest.topLeft.applying(transform), topRight: biggest.topRight.applying(transform), bottomRight: biggest.bottomRight.applying(transform), bottomLeft: biggest.bottomLeft.applying(transform))
+                    completion(biggestRectangle)
+                    
+                } else { completion(nil) }
```

Comment:
Although this class seems to be hard to test with system classes like `VNRectangleObservation`, parts of it are testable by extracting them from the code. 

You want to test for example the outcome based on the result:

- Is `quads.biggest()` returned when `results.count > 1`
- Is the transform correctly calculated?
- Is the returned `Quadrilateral` correct?
- Is the rectangleDetectionRequest called on the `userInitiated` thread?

It seems more work than it is, it should be doable!

Reply:
Cheers, I think some of the perceived difficulty is probably unfamiliarity of the code on my part. So that breakdown has really helped me :)